File: server/src/main/kotlin/xyz/luan/crusher/api/CronApi.kt
package xyz.luan.crusher.api

import com.google.common.annotations.VisibleForTesting
import com.google.gson.Gson
import spark.kotlin.Http
import spark.kotlin.RouteHandler
import spark.kotlin.halt
import xyz.luan.crusher.FirebaseWrapper
import xyz.luan.crusher.model.Cron
import xyz.luan.crusher.model.Db
import xyz.luan.crusher.model.Db.createCron

private val json = Gson()

@VisibleForTesting
fun String.extractToken() = this.replace("Bearer ([^)]*)".toRegex(), "$1")

private fun RouteHandler.parseCron() = json.fromJson(request.body(), Cron::class.java)

private fun RouteHandler.getToken(): String {
    val auth: String = request.headers("Authorization") ?: throw halt("Auth token is required")
    return auth.extractToken()
}

object CronApi {
    fun routes(http: Http) {
        http.get("/crons") {
            val email = FirebaseWrapper.validateTokenAndGetEmail(getToken())
            val crons: List<Cron> = Db.listCrons(email)
            json.toJson(crons)
        }
        http.post("/crons") {
            val email = FirebaseWrapper.validateTokenAndGetEmail(getToken())
            val cron = parseCron()
            if (cron.userEmail != email) halt(403, "You can only created crons for yourself!")
            cron.validate()
            json.toJson(createCron(cron))
        }
        http.delete("/crons/:id") {
            val email = FirebaseWrapper.validateTokenAndGetEmail(getToken())
            val id = request.params("id")?.let { Integer.parseInt(it) } ?: throw halt(404, "Invalid id provided.")
            val dbCron = Db.findDbCron(id) ?: throw halt(404, "Cron does not exist")
            if (dbCron.userEmail != email) throw halt(403, "You can only delete your own crons")
            Db.deleteCron(dbCron)
            "Removed successfully"
        }
    }
}

File: app/lib/api/cron_api.dart
import 'dart:convert' show json;
import 'package:crusher/api/api_exception.dart';
import 'package:http/http.dart' as http;

import '../login.dart';
import '../model/cron.dart';

class CronApi {
  static const BASE = 'https://crusher-app.herokuapp.com/';

  static Future<List<Cron>> fetchCrons(Login user) async {
    final response =
        await http.get('$BASE/crons', headers: await _headers(user));
    _checkStatusCode(response);
    final List list = (json.decode(response.body) as List);
    return list.map((m) => Cron.fromJson(m)).toList();
  }

  static Future<Cron> createCron(Login user, Cron cron) async {
    final String body = json.encode(cron.toMap());
    final response = await http.post('$BASE/crons',
        body: body, headers: await _headers(user));
    _checkStatusCode(response);
    return Cron.fromJson(json.decode(response.body) as Map);
  }

  static Future<void> deleteCron(Login user, Cron cron) async {
    final http.Response response = await http.delete('$BASE/crons/${cron.id}',
        headers: await _headers(user));
    _checkStatusCode(response);
  }

  static Future<Map<String, String>> _headers(Login user) async {
    final String token = await user.getToken();
    final Map<String, String> headers = {'Authorization': 'Bearer $token'};
    return headers;
  }

  static void _checkStatusCode(http.Response response) {
    if (response.statusCode != 200) throw ApiException(response);
  }
}
